File: crates/seeed-hal-client/src/gpio.rs
use std::time::Duration;

use seeed_hal_core::{ErrorCategory, HalError, HalResult, LeaseToken, ResourceId, SessionId};
use seeed_hal_gpio::{GpioEdgeEvent, GpioEdgeRequest, MAX_GPIO_EVENTS};
use seeed_hal_protocol::v1::{self, envelope};
use seeed_hal_protocol::{
    gpio_next_edge_response_from_proto, gpio_read_response_from_proto,
    open_gpio_response_from_proto,
};

use crate::HalClient;
use crate::connection::ExpectedResponse;

/// An opaque broker-owned GPIO line-group session.
#[must_use = "a remote GPIO handle owns a broker session until explicitly closed"]
pub struct RemoteGpioHandle {
    client: HalClient,
    resource_id: ResourceId,
    session_id: SessionId,
    lease: LeaseToken,
    line_count: usize,
    closed: bool,
}

impl RemoteGpioHandle {
    pub(crate) fn from_response(
        client: HalClient,
        resource_id: ResourceId,
        line_count: usize,
        response: v1::OpenGpioResponse,
    ) -> HalResult<Self> {
        let (session_id, lease) = open_gpio_response_from_proto(response)
            .map_err(|error| attach_resource(error, &resource_id))?;
        Ok(Self {
            client,
            resource_id,
            session_id,
            lease,
            line_count,
            closed: false,
        })
    }

    pub async fn read(&self) -> HalResult<Vec<bool>> {
        self.ensure_open("gpio.read")?;
        let payload = self
            .client
            .send(
                envelope::Payload::GpioReadRequest(v1::GpioReadRequest {
                    session_id: self.session_id.as_str().to_owned(),
                    lease: Some((&self.lease).into()),
                }),
                ExpectedResponse::GpioRead {
                    line_count: self.line_count,
                    resource_id: self.resource_id.clone(),
                },
            )
            .await
            .map_err(|error| attach_resource(error, &self.resource_id))?;
        let envelope::Payload::GpioReadResponse(response) = payload else {
            unreachable!()
        };
        gpio_read_response_from_proto(response)
            .and_then(|values| {
                if values.len() == self.line_count {
                    Ok(values)
                } else {
                    Err(self.local_error(
                        "runtime.protocol.invalid_message",
                        ErrorCategory::InvalidArgument,
                        "gpio.read",
                        false,
                        "GPIO read response length does not match opened lines",
                    ))
                }
            })
            .inspect_err(|error| {
                self.client.fail(error.clone());
            })
    }

    pub async fn write(&self, values: Vec<bool>) -> HalResult<()> {
        self.ensure_open("gpio.write")?;
        if values.len() != self.line_count || values.len() > MAX_GPIO_EVENTS {
            return Err(self.local_error(
                "runtime.argument.invalid",
                ErrorCategory::InvalidArgument,
                "gpio.write",
                false,
                "GPIO write values must match opened lines",
            ));
        }
        let payload = envelope::Payload::GpioWriteRequest(v1::GpioWriteRequest {
            session_id: self.session_id.as_str().to_owned(),
            lease: Some((&self.lease).into()),
            values,
        });
        self.client
            .ensure_payload_for_resource(&payload, "gpio.write", &self.resource_id)?;
        self.client
            .send(
                payload,
                ExpectedResponse::GpioWrite {
                    resource_id: self.resource_id.clone(),
                },
            )
            .await
            .map_err(|error| attach_resource(error, &self.resource_id))?;
        Ok(())
    }

    pub async fn next_edge(
        &self,
        request: GpioEdgeRequest,
        timeout: Duration,
    ) -> HalResult<Option<GpioEdgeEvent>> {
        self.ensure_open("gpio.next_edge")?;
        let timeout_ms = u64::try_from(timeout.as_millis()).map_err(|_| {
            self.local_error(
                "runtime.argument.invalid",
                ErrorCategory::InvalidArgument,
                "gpio.next_edge",
                false,
                "GPIO edge timeout exceeds the wire range",
            )
        })?;
        if timeout_ms == 0 {
            return Err(self.local_error(
                "runtime.argument.invalid",
                ErrorCategory::InvalidArgument,
                "gpio.next_edge",
                false,
                "GPIO edge timeout must be non-zero",
            ));
        }
        self.client
            .require_gpio_edges("gpio.next_edge", &self.resource_id)?;
        let payload = envelope::Payload::GpioNextEdgeRequest(v1::GpioNextEdgeRequest {
            session_id: self.session_id.as_str().to_owned(),
            lease: Some((&self.lease).into()),
            rising: request.edges().contains(seeed_hal_gpio::GpioEdge::Rising),
            falling: request.edges().contains(seeed_hal_gpio::GpioEdge::Falling),
            capacity: request.capacity() as u32,
            timeout_ms,
        });
        self.client
            .ensure_payload_for_resource(&payload, "gpio.next_edge", &self.resource_id)?;
        let payload = self
            .client
            .send(
                payload,
                ExpectedResponse::GpioNextEdge {
                    resource_id: self.resource_id.clone(),
                },
            )
            .await
            .map_err(|error| attach_resource(error, &self.resource_id))?;
        let envelope::Payload::GpioNextEdgeResponse(response) = payload else {
            unreachable!()
        };
        gpio_next_edge_response_from_proto(response).map_err(|error| {
            let error = attach_resource(error, &self.resource_id);
            self.client.fail(error.clone());
            error
        })
    }

    pub async fn close(&mut self) -> HalResult<()> {
        self.ensure_open("gpio.close")?;
        let payload = envelope::Payload::CloseGpioRequest(v1::CloseGpioRequest {
            session_id: self.session_id.as_str().to_owned(),
            lease: Some((&self.lease).into()),
        });
        self.client
            .send(
                payload,
                ExpectedResponse::CloseGpio {
                    resource_id: self.resource_id.clone(),
                },
            )
            .await
            .map_err(|error| attach_resource(error, &self.resource_id))?;
        self.closed = true;
        Ok(())
    }

    fn ensure_open(&self, operation: &'static str) -> HalResult<()> {
        if self.closed {
            Err(self.local_error(
                "runtime.session.closed",
                ErrorCategory::Conflict,
                operation,
                false,
                "the remote GPIO handle is closed",
            ))
        } else {
            Ok(())
        }
    }
    fn local_error(
        &self,
        name: &'static str,
        category: ErrorCategory,
        operation: &'static str,
        retryable: bool,
        message: &'static str,
    ) -> HalError {
        HalError::new(name, category, operation, retryable, message)
            .expect("static GPIO client error metadata is valid")
            .with_resource_id(self.resource_id.clone())
    }
}

fn attach_resource(error: HalError, resource_id: &ResourceId) -> HalError {
    if error.resource_id().is_some() {
        error
    } else {
        error.with_resource_id(resource_id.clone())
    }
}

File: bindings/python/robot_hal/gpio.py
"""Immutable, protobuf-independent GPIO values and broker session."""

from __future__ import annotations

from dataclasses import dataclass
from enum import Enum
from typing import TYPE_CHECKING

from .errors import ErrorCategory, HalError, client_error

if TYPE_CHECKING:
    from .client import HalClient


MAX_GPIO_EVENTS = 1024


class GpioBias(Enum):
    DISABLED = "disabled"
    PULL_UP = "pull_up"
    PULL_DOWN = "pull_down"


class GpioDrive(Enum):
    PUSH_PULL = "push_pull"
    OPEN_DRAIN = "open_drain"
    OPEN_SOURCE = "open_source"


class GpioDirection(Enum):
    INPUT = "input"
    OUTPUT = "output"


class GpioEdge(Enum):
    RISING = "rising"
    FALLING = "falling"


@dataclass(frozen=True, slots=True)
class GpioLineConfig:
    direction: GpioDirection
    active_low: bool = False
    bias: GpioBias = GpioBias.DISABLED
    drive: GpioDrive | None = None
    initial_value: bool | None = None

    def __post_init__(self) -> None:
        if not isinstance(self.direction, GpioDirection) or not isinstance(self.active_low, bool) or not isinstance(self.bias, GpioBias):
            raise _invalid("GPIO configuration is invalid")
        output = self.direction is GpioDirection.OUTPUT
        if output != (self.drive is not None and self.initial_value is not None):
            raise _invalid("GPIO output requires drive and initial value")
        if self.drive is not None and not isinstance(self.drive, GpioDrive):
            raise _invalid("GPIO drive is invalid")
        if self.initial_value is not None and not isinstance(self.initial_value, bool):
            raise _invalid("GPIO initial value is invalid")


@dataclass(frozen=True, slots=True)
class GpioEdgeRequest:
    rising: bool
    falling: bool
    capacity: int

    def __post_init__(self) -> None:
        if not isinstance(self.rising, bool) or not isinstance(self.falling, bool) or not (self.rising or self.falling) or not isinstance(self.capacity, int) or isinstance(self.capacity, bool) or not 0 < self.capacity <= MAX_GPIO_EVENTS:
            raise _invalid("GPIO edge request is invalid")


@dataclass(frozen=True, slots=True)
class GpioEdgeEvent:
    edge: GpioEdge
    monotonic_ns: int
    sequence: int


class GpioSession:
    __slots__ = ("_client", "_session_id", "_lease_id", "_generation", "_mode", "_resource_id", "_line_count", "_closed")

    def __init__(self, client: HalClient, session_id: str, lease_id: str, generation: int, mode: int, resource_id: str, line_count: int) -> None:
        self._client, self._session_id, self._lease_id = client, session_id, lease_id
        self._generation, self._mode, self._resource_id, self._line_count, self._closed = generation, mode, resource_id, line_count, False

    async def read(self) -> tuple[bool, ...]:
        self._ensure_open("gpio.read")
        return await self._client._gpio_read(self)

    async def write(self, values: tuple[bool, ...]) -> None:
        self._ensure_open("gpio.write")
        await self._client._gpio_write(self, values)

    async def next_edge(self, request: GpioEdgeRequest, timeout_ms: int) -> GpioEdgeEvent | None:
        self._ensure_open("gpio.next_edge")
        return await self._client._gpio_next_edge(self, request, timeout_ms)

    async def close(self) -> None:
        if not self._closed:
            await self._client._gpio_close(self)
            self._closed = True

    def _ensure_open(self, operation: str) -> None:
        if self._closed:
            raise HalError("runtime.session.closed", ErrorCategory.CONFLICT, operation, False, "GPIO session is closed", resource_id=self._resource_id)


def _invalid(message: str) -> HalError:
    return client_error("gpio.configuration.invalid", ErrorCategory.INVALID_ARGUMENT, "gpio.configuration", False, message)
